include/mfmg/utils.cuh

```diff
@@ -0,0 +1,146 @@
+/*************************************************************************
+ * Copyright (c) 2017 by the mfmg authors                                *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef UTILS_CUH
+#define UTILS_CUH
+
+#ifdef MFMG_WITH_CUDA
+
+#include <mfmg/exceptions.hpp>
+
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/trilinos_index_access.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+
+#include <tuple>
+
+namespace mfmg
+{
+namespace internal
+{
+template <typename ScalarType>
+std::tuple<ScalarType *, int *, int *>
+copy_to_gpu(std::vector<ScalarType> const &val,
+            std::vector<int> const &column_index,
+            std::vector<int> const &row_ptr)
+{
+  unsigned int const nnz = val.size();
+  unsigned int const row_ptr_size = row_ptr.size();
+
+  // Copy the elements to the gpu
+  ScalarType *val_dev;
+  cudaError_t error_code = cudaMalloc(&val_dev, nnz * sizeof(ScalarType));
+  ASSERT_CUDA(error_code);
+  error_code = cudaMemcpy(val_dev, &val[0], nnz * sizeof(ScalarType),
+                          cudaMemcpyHostToDevice);
+  ASSERT_CUDA(error_code);
+
+  // Copy the column indices to the gpu
+  int *column_index_dev;
+  error_code = cudaMalloc(&column_index_dev, nnz * sizeof(int));
+  ASSERT_CUDA(error_code);
+  error_code = cudaMemcpy(column_index_dev, &column_index[0], nnz * sizeof(int),
+                          cudaMemcpyHostToDevice);
+  ASSERT_CUDA(error_code);
+
+  // Copy the row pointer to the gpu
+  int *row_ptr_dev;
+  error_code = cudaMalloc(&row_ptr_dev, row_ptr_size * sizeof(int));
+  ASSERT_CUDA(error_code);
+  error_code = cudaMemcpy(row_ptr_dev, &row_ptr[0], row_ptr_size * sizeof(int),
+                          cudaMemcpyHostToDevice);
+  ASSERT_CUDA(error_code);
+
+  return std::make_tuple(val_dev, column_index_dev, row_ptr_dev);
+}
+}
+
+/**
+ * Convert a dealii::SparseMatrix to the regular CSR format and move the data to
+ * the GPU.
+ */
+template <typename ScalarType>
+std::tuple<ScalarType *, int *, int *>
```

Comment:
How about creating a struct or something similar and using that instead of tuple? That would make it clear that it is a device matrix.

Reply:
OK I'll change it.